src/main/java/com/epam/wl/dao/BookDAO.java

```diff
@@ -0,0 +1,143 @@
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+
+        return bookInstances;
+    }
+
+    public int getBookId(String author, String title, int year) {
+        int id = -1;
+
+        try (Connection connection = dataSource.getConnection()) {
+            final PreparedStatement preparedStatement = connection.prepareStatement(
+                    "SELECT * FROM book WHERE author=? AND title=? AND year=?");
+            preparedStatement.setString(1, author);
+            preparedStatement.setString(2, title);
+            preparedStatement.setInt(3, year);
+            ResultSet resultSet = preparedStatement.executeQuery();
+
+            if (resultSet.first()) {
+                id = resultSet.getInt("id");
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+
+        return id;
+    }
+
+    public int getFreeBookInstanceId(int bookId) {
+        int id = -1;
+
+        try (Connection connection = dataSource.getConnection()) {
+            final Statement statement = connection.createStatement();
+            final ResultSet bookInstancesSet = statement.executeQuery(
+                    "SELECT * FROM book_instance WHERE bookid=" + bookId);
+
+            List<Integer> bookInstancesId = new ArrayList<>();
+
+            while (bookInstancesSet.next()) {
+                bookInstancesId.add(bookInstancesSet.getInt("id"));
+            }
+
+            final Statement secondStatement = connection.createStatement();
+            final ResultSet bookOrdersSet = secondStatement.executeQuery("SELECT * FROM book_order");
+
+            while (bookOrdersSet.next()) {
+                if (bookInstancesId.contains(bookOrdersSet.getInt("book_instanceid"))) {
+                    bookInstancesId.remove((Integer) bookOrdersSet.getInt("book_instanceid"));
+                }
+            }
+
+            Collections.sort(bookInstancesId); // to sort or not to sort? seems like not
+
+            if (!bookInstancesId.isEmpty()) {
+                id = bookInstancesId.get(0);
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+
+        return id;
+    }
+
+    public void addNewBook(String author, String title, int year) {
+        try (Connection connection = dataSource.getConnection()) {
+            final PreparedStatement preparedStatement = connection.prepareStatement(
+                    "INSERT INTO book(author, title, year) VALUES (?, ?, ?)");
+            preparedStatement.setString(1, author);
+            preparedStatement.setString(2, title);
+            preparedStatement.setInt(3, year);
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void addNewBookInstance(String author, String title, int year) {
+        try (Connection connection = dataSource.getConnection()) {
```

Comment:
Can these two methods be combined into one? If you're a librarian, you'd just want to add a book that already exist. There's no point in adding a book without an instance of it.